### PR TITLE
More tests for migrate action

### PR DESCRIPTION
**Pull request by karuppayya**
- Adding more tests for migrate action

### CODE REVIEW

**Comment by karuppayya** [comment]
Tagging @RussellSpitzer @aokolnychyi @rdblue for review. Thanks!

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 1]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+    Schema beforeSchema = table.schema();
+    String colName = "newCol";
+    sparkTable.table().updateSchema().addColumn("newCol", Types.IntegerType.get()).commit();
+    Schema afterSchema = table.schema();
+    Assert.assertNull(beforeSchema.findField(colName));
+    Assert.assertNotNull(afterSchema.findField(colName));
+    // reads should succeed without any exceptions
+    spark.table(dest).collect();
+
+    colName = "newCol1";
```
Comment (RussellSpitzer) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:228]: nit: I would have this be a different variable rather than reusing "colName"

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 2]
```diff
@@ -296,14 +406,25 @@ private void createSourceTable(String createStatement, String tableName)
...
+      uris.add(sourceTable.location());
+    } else {
+      Seq<CatalogTablePartition> catalogTablePartitionSeq =
+              spark.sessionState().catalog().listPartitions(sourceTable.identifier(), Option.apply(null));
+      uris = JavaConverters.seqAsJavaList(catalogTablePartitionSeq)
+              .stream()
+              .map(CatalogTablePartition::location)
+              .collect(Collectors.toList());
+    }
+    long expectedMigratedFiles = uris.stream()
```
Comment (RussellSpitzer) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:421]: Is this change needed? Is this for checking if we have partitions that aren't in the sourcetable directory?

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 3]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
     assertMigratedFileCount(Actions.migrate(source), source, dest);
   }
 
+  @Test
+  public void testPartitionedTableWithUnRecoveredPartitions() throws Exception {
```
Comment (RussellSpitzer) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:172]: What is this testing? I'm not familiar with what unrecovered partitions are.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 4]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+    String source = sourceName("test_custom_parts");
+    String dest = source;
+    File tblLocation = temp.newFolder();
+    File partitionDataLoc = temp.newFolder();
+    spark.sql(String.format(CREATE_PARTITIONED_PARQUET, source, tblLocation));
+    spark.range(10)
+            .selectExpr("cast(id as STRING) as data")
+            .write()
+            .mode("overwrite").parquet(partitionDataLoc.toURI().toString());
+    spark.sql(String.format("alter table %s add partition(id=0) LOCATION '%s'", source,
```
Comment (RussellSpitzer) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:201]: This is why we modify the file lookup algorithm? Because we have to look up a partition with a custom location?

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 3 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:172]: Example usecase:
There is  a location where a external system is writing files every day.
An external partitioned Hive table is created on top of the location, partitioned on day.
Once the external table is created, we need to run "ALTER TABLE RECOVER PARTITIONS" or "MSCK REPAIR TABLE" to register the partitions in metastore. .

When we dont recover(register against metastore) a partitions(intentionally or unintentionally), the data corresponding to unrecovered partition is not returned as part of the table scan.
This test checks if both Spark and Iceberg returns same data, when there are unrecovered partitions

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 2 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:421]: For the tests regarding unrecovered partitions(one of the new test case), the total file count should be that of only those those registered in metastore. With the prior change, it does a root level listing, counting files from unrecovered partitions as well.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 4 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:201]: Partly yes. The change was also required for the another test case regarding unrecovered partitions.

**Comment by RussellSpitzer** [comment]
These all look like good tests to me :+1:

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 3 above)
Comment (aokolnychyi) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:172]: Sounds like a good thing to test.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 5]
```diff
@@ -67,6 +76,8 @@
       "PARTITIONED BY (id INT) STORED AS parquet LOCATION '%s'";
   private static final String CREATE_HIVE_PARQUET = "CREATE TABLE %s (data STRING) " +
       "PARTITIONED BY (id INT) STORED AS parquet";
+  private static final String CREATE_HIVE_PARQUET_UNPARTITIONED = "CREATE EXTERNAL TABLE %s (id INT, data STRING) " +
+          "STORED AS parquet LOCATION '%s'";
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:80]: Iceberg uses 2 spaces for regular formatting and 4 spaces for continued lines (like above). This line should be 4 instead of 8.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 6]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+  public void testPartitionedTableWithUnRecoveredPartitions() throws Exception {
+    Assume.assumeTrue("Cannot migrate to a hadoop based catalog", !type.equals("hadoop"));
+    Assume.assumeTrue("Can only migrate from Spark Session Catalog", catalog.name().equals("spark_catalog"));
+    String source = sourceName("test_unrecovered_partitions");
+    String dest = source;
+    File location = temp.newFolder();
+    spark.sql(String.format(CREATE_PARTITIONED_PARQUET, source, location));
+    spark.range(5)
+            .selectExpr("id", "cast(id as STRING) as data")
+            .write()
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:181]: These lines should be formatted with 4 spaces.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 7]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
   }
 
+  @Test
+  public void testPartitionedTableWithUnRecoveredPartitions() throws Exception {
+    Assume.assumeTrue("Cannot migrate to a hadoop based catalog", !type.equals("hadoop"));
+    Assume.assumeTrue("Can only migrate from Spark Session Catalog", catalog.name().equals("spark_catalog"));
+    String source = sourceName("test_unrecovered_partitions");
+    String dest = source;
+    File location = temp.newFolder();
+    spark.sql(String.format(CREATE_PARTITIONED_PARQUET, source, location));
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:178]: Can we use `sql` method defined in tests instead of `String.format`?

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 8]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+    spark.range(5)
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:179]: nit: I usually prefer to have an empty line between different logical blocks. That way, it is easier to read.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 9]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+    Assume.assumeTrue("Cannot migrate to a hadoop based catalog", !type.equals("hadoop"));
+    Assume.assumeTrue("Can only migrate from Spark Session Catalog", catalog.name().equals("spark_catalog"));
+    String source = sourceName("test_custom_parts");
+    String dest = source;
+    File tblLocation = temp.newFolder();
+    File partitionDataLoc = temp.newFolder();
+    spark.sql(String.format(CREATE_PARTITIONED_PARQUET, source, tblLocation));
+    spark.range(10)
+            .selectExpr("cast(id as STRING) as data")
+            .write()
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:199]: nit: formatting

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 4 above)
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:201]: spark.sql -> sql

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 10]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+    Assume.assumeTrue("Can only migrate from Spark Session Catalog", catalog.name().equals("spark_catalog"));
+    String source = sourceName("test_add_column_migrated_table");
+    String dest = source;
+    createSourceTable(CREATE_PARQUET, source);
+    Actions.migrate(source).execute();
+    SparkTable sparkTable = loadTable(dest);
+    Table table = sparkTable.table();
+
+    // test column addition on migrated table
+    Schema beforeSchema = table.schema();
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:219]: Is there a particular reason not to use SQL statements for this?

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 11]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+    String colName = "newCol";
+    sparkTable.table().updateSchema().addColumn("newCol", Types.IntegerType.get()).commit();
+    Schema afterSchema = table.schema();
+    Assert.assertNull(beforeSchema.findField(colName));
+    Assert.assertNotNull(afterSchema.findField(colName));
+    // reads should succeed without any exceptions
+    spark.table(dest).collect();
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:226]: Isn't the table empty at this step? Shall we validate the output too?

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 12]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+            .write()
+            .mode("overwrite").parquet(partitionDataLoc.toURI().toString());
+    spark.sql(String.format("alter table %s add partition(id=0) LOCATION '%s'", source,
+            partitionDataLoc.toURI().toString()));
+    assertMigratedFileCount(Actions.migrate(source), source, dest);
+  }
+
+
+  @Test
+  public void testAddColumnOnMigratedTable() throws Exception {
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:208]: It would be great to add columns not only to the end but also in the middle. That's a trickier change as the built-in file sources don't support that. Name mapping in Iceberg should cover this use case but we have to verify it.

Comment (aokolnychyi) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:208]: @karuppayya, you can probably just add more column changes to this test.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 13]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+    Assert.assertNotNull(afterSchema.findField(colName));
+    // reads should succeed without any exceptions
+    spark.table(dest).collect();
+
+    colName = "newCol1";
+    spark.sql(String.format("ALTER TABLE %s ADD COLUMN %s INT", dest, colName)).collect();
+    StructType schema = spark.table(dest).schema();
+    Assert.assertTrue(Arrays.asList(schema.fieldNames()).contains(colName));
+    // reads should succeed without any exceptions
+    spark.table(dest).collect();
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:233]: Should validate the output too.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 14]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+  public void testRemoveColumnOnMigratedTable() throws Exception {
+    Assume.assumeTrue("Cannot migrate to a hadoop based catalog", !type.equals("hadoop"));
+    Assume.assumeTrue("Can only migrate from Spark Session Catalog", catalog.name().equals("spark_catalog"));
+    String source = sourceName("test_remove_column_migrated_table");
+    String dest = source;
+    File location = temp.newFolder();
+    String colName1 = "newCol1";
+    String colName2 = "newCol2";
+    String createStmt = "CREATE TABLE %s (id INT, data STRING, %s INT, %s INT) using parquet LOCATION '%s'";
+    spark.sql(String.format(createStmt, source, colName1, colName2, location));
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:247]: spark.sql -> sql

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 10 above)
Comment (aokolnychyi) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:219]: Oh, we use both the API and SQL. Good!

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 15]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+    Table table = sparkTable.table();
+
+    // test column removal on migrated table
+    Schema beforeSchema = table.schema();
+    sparkTable.table().updateSchema().deleteColumn(colName1).commit();
+    Schema afterSchema = table.schema();
+    Assert.assertNotNull(beforeSchema.findField(colName1));
+    Assert.assertNull(afterSchema.findField(colName1));
+    // reads should succeed without any exceptions
+    spark.table(dest).collect();
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:261]: Should validate the output.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 16]
```diff
@@ -157,6 +168,105 @@ public void testMigratePartitioned() throws Exception {
...
+
+    spark.sql(String.format("ALTER TABLE %s DROP COLUMN %s", dest, colName2)).collect();
+    StructType schema = spark.table(dest).schema();
+    Assert.assertFalse(Arrays.asList(schema.fieldNames()).contains(colName2));
+    // reads should succeed without any exceptions
+    spark.table(dest).collect();
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:267]: Should validate the output.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 17]
```diff
@@ -296,14 +406,25 @@ private void createSourceTable(String createStatement, String tableName)
...
-        .collect(Collectors.toList());
-    long expectedMigratedFiles = files.size();
-    List<Row> expected = spark.table(source).collectAsList();
+    List<URI> uris;
+    if (sourceTable.partitionColumnNames().size() == 0) {
+      uris = new ArrayList<>();
+      uris.add(sourceTable.location());
+    } else {
+      Seq<CatalogTablePartition> catalogTablePartitionSeq =
+              spark.sessionState().catalog().listPartitions(sourceTable.identifier(), Option.apply(null));
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:415]: should be formatted with 4 spaces instead of 8.

**Comment by aokolnychyi** [comment]
Thanks for working on this, @karuppayya! I feel these tests would be helpful. I did one pass and left some minor comments.

In addition, it would be great to add one or more tests when the source table has an evolved schema. I believe the only operation we can perform in the built-in file sources is to add columns at the end. Can we add a test when a table is modified before and after migration?

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 1 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java]: done

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 5 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java]: Thanksfor lettting me know!!, I have fixed the indentations.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 6 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java]: done

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 7 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java]: done

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 9 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java]: fixed

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 11 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java]: Added output correctness checks

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 17 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java]: Fixed

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 18]
```diff
@@ -70,6 +79,10 @@
 
   private static final String NAMESPACE = "default";
 
+  // Error message
+  private static final String OUTPUT_DIFFERENT_ERROR_MSG =
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:83]: nit: I think it is alright to use the error message in place. To stay on one line, we can simplify the message like `Output must match` or anything appropriate.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 19]
```diff
@@ -157,6 +170,192 @@ public void testMigratePartitioned() throws Exception {
...
+  public void testPartitionedTableWithCustomPartitions() throws Exception {
+    Assume.assumeTrue("Cannot migrate to a hadoop based catalog", !type.equals("hadoop"));
+    Assume.assumeTrue("Can only migrate from Spark Session Catalog", catalog.name().equals("spark_catalog"));
+    String source = sourceName("test_custom_parts");
+    String dest = source;
+    File tblLocation = temp.newFolder();
+    File partitionDataLoc = temp.newFolder();
+
+    // Data generation and partition addition
+    spark.sql(String.format(CREATE_PARTITIONED_PARQUET, source, tblLocation));
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:203]: nit: spark.sql -> sql

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 20]
```diff
@@ -157,6 +170,192 @@ public void testMigratePartitioned() throws Exception {
...
+    spark.range(10)
+        .selectExpr("cast(id as STRING) as data")
+        .write()
+        .mode(SaveMode.Overwrite).parquet(partitionDataLoc.toURI().toString());
+    spark.sql(String.format("ALTER TABLE %s ADD PARTITION(id=0) LOCATION '%s'", source,
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:208]: nit: spark.sql -> sql

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 12 above)
Comment (aokolnychyi) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:214]: Since methods are independent, I think we better have two tests here. What do you think, @karuppayya?

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 21]
```diff
@@ -157,6 +170,192 @@ public void testMigratePartitioned() throws Exception {
...
+    Assume.assumeTrue("Cannot migrate to a hadoop based catalog", !type.equals("hadoop"));
+    Assume.assumeTrue("Can only migrate from Spark Session Catalog", catalog.name().equals("spark_catalog"));
+    String source = sourceName("test_add_column_migrated_table");
+    String dest = source;
+    createSourceTable(CREATE_PARQUET, source);
+    List<Object[]> expected1 = sql("select *, null from %s order by id", dest);
+    List<Object[]> expected2 = sql("select *, null, null from %s order by id", dest);
+
+    // migrate table
+    Actions.migrate(source).execute();
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:229]: @karuppayya, am I correct that the source table is empty when we call the migrate action?

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 22]
```diff
@@ -157,6 +170,192 @@ public void testMigratePartitioned() throws Exception {
...
+
+    // reads should succeed
+    List<Object[]> results = sql("select * from %s order by id", dest);
+    Assert.assertTrue(results.size() > 0);
+    assertEquals(OUTPUT_DIFFERENT_ERROR_MSG, results, expected);
+  }
+
+  @Test
+  public void testRemoveColumnOnMigratedTable() throws Exception {
+    removeColumnsAtEnd();
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:288]: Same here: deserves two tests

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 23]
```diff
@@ -157,6 +170,192 @@ public void testMigratePartitioned() throws Exception {
...
+
+  private void testAddColumnOnMigratedTableAtMiddle() throws Exception {
+    Assume.assumeTrue("Cannot migrate to a hadoop based catalog", !type.equals("hadoop"));
+    Assume.assumeTrue("Can only migrate from Spark Session Catalog", catalog.name().equals("spark_catalog"));
+    String source = sourceName("test_add_column_migrated_table_middle");
+    String dest = source;
+    createSourceTable(CREATE_PARQUET, source);
+
+    // migrate table
+    Actions.migrate(source).execute();
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:265]: Same here: the source seems empty

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 24]
```diff
@@ -157,6 +170,192 @@ public void testMigratePartitioned() throws Exception {
...
+    testAddColumnOnMigratedTableAtMiddle();
+  }
+
+  private void testAddColumnOnMigratedTableAtEnd() throws Exception {
+    Assume.assumeTrue("Cannot migrate to a hadoop based catalog", !type.equals("hadoop"));
+    Assume.assumeTrue("Can only migrate from Spark Session Catalog", catalog.name().equals("spark_catalog"));
+    String source = sourceName("test_add_column_migrated_table");
+    String dest = source;
+    createSourceTable(CREATE_PARQUET, source);
+    List<Object[]> expected1 = sql("select *, null from %s order by id", dest);
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:225]: I think it would be more readable if this queried `source` even though they are identical. As we are trying to ensure the output before and after migration match.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java [hunk 25]
```diff
@@ -157,6 +170,192 @@ public void testMigratePartitioned() throws Exception {
...
+    String source = sourceName("test_remove_column_migrated_table");
+    String dest = source;
+
+    String colName1 = "newCol1";
+    String colName2 = "newCol2";
+    File location = temp.newFolder();
+    spark.range(10).selectExpr("cast(id as INT)", "CAST(id as INT) " + colName1, "CAST(id as INT) " + colName2)
+        .write()
+        .mode(SaveMode.Overwrite).saveAsTable(dest);
+    List<Object[]> expected1 = sql("select id, %s from %s order by id", colName1, dest);
```
Comment (aokolnychyi) [inline spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:304]: Same here: we better query `source` so that it is easier to follow.

**Comment by aokolnychyi** [comment]
@karuppayya, looks like we don't have a test when the source table has an evolved schema before the migration. Did I miss it or do you plan to add one in a follow-up?

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 21 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:229]: `createSourceTable(CREATE_PARQUET, source)` line#224  has the logic to populate the data for the table.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 19 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:199]: done

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 23 above)
Comment (karuppayya) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:256]: `createSourceTable` populates data

**Comment by karuppayya** [comment]
@aokolnychyi Thanks for reviewing. I have added couple of tests covering the schema evolution cases.

**Comment by rdblue** [comment]
@aokolnychyi and @RussellSpitzer, can you take another look? Is this ready to merge?

**Comment by aokolnychyi** [comment]
I will do another pass today.

File: spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java (hunk 21 above)
Comment (aokolnychyi) [reply spark3/src/test/java/org/apache/iceberg/actions/TestCreateActions.java:219]: Got it, I missed it.

**Review by aokolnychyi — APPROVED** [review]
This change looks good to me. Any other comments, @RussellSpitzer?

**Comment by aokolnychyi** [comment]
I just realized Russell had already given his approval 15 days ago. Since it was open for a while and all comments seem to be resolved, I'll merge it.

Thanks, @karuppayya!

**Comment by RussellSpitzer** [comment]
Ah sorry! I need to fix my alert notifications
